community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/TemporalIndexCache.java

```diff
@@ -157,95 +155,83 @@ T select( ValueGroup valueGroup ) throws IOException
         return cachedValue != null ? function.apply( cachedValue ) : orElse;
     }
 
-    private T date() throws UncheckedIOException
+    private void acquireInstantiateCloseLock()
     {
-        return cache.computeIfAbsent( date, d ->
+        instantiateCloseLock.lock();
+        if ( closed )
         {
-            try
-            {
-                return factory.newDate();
-            }
-            catch ( IOException e )
-            {
-                throw new UncheckedIOException( e );
-            }
-        } );
+            instantiateCloseLock.unlock();
+            throw new IllegalStateException( this + " is already closed" );
```

Comment:
I have two concerns about this method:

1. It feels a bit weird to do the close check when we actually just want to grab the lock and it creates some unexpected behaviour. We will not be able to call close twice for example and I think close should be an idempotent operation.
How about checking the closed flag under the lock explicitly in `getOrCreatePart` and act accordingly there? 

2. This exception will be thrown in the same scenario where we previously leaked of file mappings, correct? I have some trouble seeing what implication that has. I hope it will not:
a. ... cause kernel panic (that would be no better that current situation). 
b. ... fail any user transactions.
c. ... be thrown at users.

Reply:
(1): 👍 I like your suggestion better
(2): think of a number index, calling `newUpdater` after `close()` results in `IllegalStateException`, this is the same thing, isn't it?